### PR TITLE
v1.0 android

### PR DESCRIPTION
**Pull request by ErisMik**

### CODE REVIEW

File: binding/android/README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+Koala requires a valid Picovoice `AccessKey` at initialization. `AccessKey` acts as your credentials when using Koala SDKs.
+You can get your `AccessKey` for free. Make sure to keep your `AccessKey` secret.
+Signup or Login to [Picovoice Console](https://console.picovoice.ai/) to get your `AccessKey`.
+
+## Usage
+
+Add the Koala model file to your Android application by:
+
+1. Either create a model in [Picovoice Console](https://console.picovoice.ai/) or use the [default model](../../lib/common).
```
Comment (ksyeo1010) [inline binding/android/README.md:43]: can't create models

File: binding/android/build.gradle [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+buildscript {
+    repositories {
+        maven { url "https://plugins.gradle.org/m2/" }
+        google()
+        mavenCentral()
+    }
+    dependencies {
+        classpath "com.android.tools.build:gradle:4.2.2"
+        classpath "com.github.dcendents:android-maven-gradle-plugin:2.1"
```
Comment (ksyeo1010) [inline binding/android/build.gradle:13]: this is not required now: take a look at this PR https://github.com/Picovoice/cobra/pull/133

File: binding/android/koala/src/main/java/ai/picovoice/koala/Koala.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        return KoalaNative.delaySample(handle);
+    }
+
+    public static class Builder {
+
+        private String accessKey = null;
+        private String modelPath = null;
+
+        /**
+         * Setter the AccessKey
```
Comment (ksyeo1010) [inline binding/android/koala/src/main/java/ai/picovoice/koala/Koala.java:175]:
```suggestion
         * Setter for the AccessKey
```

File: demo/android/Activity/koala-activity-demo-app/src/androidTest/java/ai/picovoice/koalaactivitydemo/KoalaTest.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
+/*
+    Copyright 2021 Picovoice Inc.
```
Comment (ksyeo1010) [inline demo/android/Activity/koala-activity-demo-app/src/androidTest/java/ai/picovoice/koalaactivitydemo/KoalaTest.java:2]:
```suggestion
    Copyright 2023 Picovoice Inc.
```

File: demo/android/Activity/koala-activity-demo-app/src/main/java/ai/picovoice/koalaactivitydemo/MainActivity.java [hunk 5]
```diff
@@ -0,0 +1,379 @@
+/*
+    Copyright 2021 Picovoice Inc.
```
Comment (ksyeo1010) [inline demo/android/Activity/koala-activity-demo-app/src/main/java/ai/picovoice/koalaactivitydemo/MainActivity.java:2]:
```suggestion
    Copyright 2023 Picovoice Inc.
```

File: binding/android/koala/consumer-rules.pro [hunk 6]
```diff
@@ -0,0 +1,2 @@
+-keep class ai.picovoice.koala.*Exception { <init>(...); }
+-keep class ai.picovoice.koala.KoalaTranscript
```
Comment (ksyeo1010) [inline binding/android/koala/consumer-rules.pro:2]:
```suggestion
```

File: binding/android/koala/proguard-rules.pro [hunk 7]
```diff
@@ -0,0 +1,2 @@
+-keep class ai.picovoice.koala.*Exception { <init>(...); }
+-keep class ai.picovoice.koala.KoalaTranscript
```
Comment (ksyeo1010) [inline binding/android/koala/proguard-rules.pro:2]:
```suggestion
```

**Review by ksyeo1010 — CHANGES_REQUESTED** [review]

File: binding/android/koala/src/main/java/ai/picovoice/koala/KoalaNative.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    static native long init(String accessKey, String modelPath) throws KoalaException;
+
+    static native void delete(long object);
+
+    static native short[] process(long object, short[] pcm) throws KoalaException;
+
+    static native void reset(long object) throws KoalaException;
+
+    static native int delaySample(long object) throws KoalaException;
```
Comment (laves) [inline binding/android/koala/src/main/java/ai/picovoice/koala/KoalaNative.java:31]: getDelaySample to follow the pattern of the property getter names

File: binding/android/settings.gradle [hunk 9]
```diff
@@ -0,0 +1,2 @@
+rootProject.name = "Koala"
```
Comment (laves) [inline binding/android/settings.gradle:1]: `libpv_koala.so` and `koala_params.pv` can be ignored and added to the gradle build. That way we don't need to remember to update them.

File: demo/android/Activity/README.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+```console
+pvTestingAccessKey=YOUR_ACCESS_KEY_HERE
+```
+
+## Usage
+
+Launch the demo on your phone using Android Studio.
+
+1. Press the start button
+2. Start talking. The app background will change to indicate that voice activity was detected above the threshold.
```
Comment (laves) [inline demo/android/Activity/README.md:27]: This doesn't describe the demo

File: demo/android/Activity/koala-activity-demo-app/src/androidTest/java/ai/picovoice/koalaactivitydemo/PerformanceTest.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        extractAssetsRecursively("test_resources");
+        testResourcesPath = new File(appContext.getFilesDir(), "test_resources").getAbsolutePath();
+
+        accessKey = appContext.getString(R.string.pvTestingAccessKey);
+    }
+
+
+
+    @Test
+    public void testPerformance() throws Exception {
```
Comment (laves) [inline demo/android/Activity/koala-activity-demo-app/src/androidTest/java/ai/picovoice/koalaactivitydemo/PerformanceTest.java:64]: formatting

File: demo/android/README.md [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+Koala requires a valid Picovoice `AccessKey` at initialization. `AccessKey` acts as your credentials when using Koala SDKs.
+You can get your `AccessKey` for free. Make sure to keep your `AccessKey` secret.
+Signup or Login to [Picovoice Console](https://console.picovoice.ai/) to get your `AccessKey`.
+
+## Activity
+
+Copy your AccessKey into the `ACCESS_KEY` variable in `MainActivity.java` before building the demo.
+
+This demo is intended for applications that need to do voice activity detection when in focus.
```
Comment (laves) [inline demo/android/README.md:13]: description from Cobra

File: demo/android/Activity/koala-activity-demo-app/src/main/res/layout/koala_activity_demo.xml [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="16dp"
+            android:onClick="onClickPlay"
+            android:text="PLAY/STOP"
+            android:textOff="PLAY"
+            android:textOn="STOP"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@+id/faderSlider" />
```
Comment (laves) [inline demo/android/Activity/koala-activity-demo-app/src/main/res/layout/koala_activity_demo.xml:66]: Can you make this Picovoice blue?

File: demo/android/Activity/koala-activity-demo-app/src/main/res/layout/koala_activity_demo.xml [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    </androidx.constraintlayout.widget.ConstraintLayout>
+
+    <TextView
+        android:id="@+id/recordedText"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginBottom="16dp"
+        android:text="Recorded"
```
Comment (laves) [inline demo/android/Activity/koala-activity-demo-app/src/main/res/layout/koala_activity_demo.xml:74]: Should only say `Recorded` after recording. During it should say `Recording`

File: demo/android/Activity/koala-activity-demo-app/src/main/res/layout/koala_activity_demo.xml [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        android:layout_height="100dp"
+        android:layout_marginStart="8dp"
+        android:layout_marginEnd="8dp"
+        android:layout_marginBottom="20dp"
+        android:background="@drawable/button_background"
+        android:onClick="onClickRecord"
+        android:textColor="@android:color/white"
+        android:textOff="Record"
+        android:textOn="Stop"
+        android:textSize="24sp"
```
Comment (laves) [inline demo/android/Activity/koala-activity-demo-app/src/main/res/layout/koala_activity_demo.xml:91]: Might be good to use a smaller text size - there's almost no padding when the text is Record

File: demo/android/Activity/koala-activity-demo-app/src/main/java/ai/picovoice/koalaactivitydemo/MainActivity.java [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+                        if (totalSamplesWritten >= koalaDelay) {
+                            writeFrame(enhancedFile, frameBufferEnhanced);
+                            enhancedSamplesWritten += frameBufferEnhanced.length;
+                        }
+                    }
+
+                    if ((totalSamplesWritten / koala.getFrameLength()) % 10 == 0) {
+                        runOnUiThread(() -> {
+                            int secondsRecorded = (totalSamplesWritten / koala.getSampleRate());
+                            recordedText.setText("Recorded: " + secondsRecorded + "s");
```
Comment (laves) [inline demo/android/Activity/koala-activity-demo-app/src/main/java/ai/picovoice/koalaactivitydemo/MainActivity.java:323]: can we do a float to one decimal place for recording time? Looks more familiar as a recording timer with some milliseconds

**Review by laves — CHANGES_REQUESTED** [review]

File: binding/android/koala/src/main/java/ai/picovoice/koala/KoalaNative.java (hunk 8 above)
Comment (laves) [reply binding/android/koala/src/main/java/ai/picovoice/koala/KoalaNative.java:31]: sure

**Comment by ErisMik** [comment]
Markdown links failing because cocoapods.org is down right now :grimacing:

**Review by ksyeo1010 — APPROVED** [review]

**Review by laves — APPROVED** [review]

**Pull request merged**